core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java

```diff
@@ -124,11 +137,101 @@ public void delete(Delete delete) throws CrudException {
     }
   }
 
+  /** @deprecated As of release 3.13.0. Will be removed in release 5.0.0. */
+  @SuppressWarnings("InlineMeSuggester")
+  @Deprecated
   @Override
   public void delete(List<Delete> deletes) throws CrudException {
     mutate(deletes);
   }
 
+  @Override
+  public void insert(Insert insert) throws CrudException {
+    insert = copyAndSetTargetToIfNot(insert);
+
+    PutBuilder.Buildable buildable =
+        Put.newBuilder()
+            .namespace(insert.forNamespace().orElse(null))
+            .table(insert.forTable().orElse(null))
+            .partitionKey(insert.getPartitionKey());
+    insert.getClusteringKey().ifPresent(buildable::clusteringKey);
+    insert.getColumns().values().forEach(buildable::value);
+    buildable.condition(ConditionBuilder.putIfNotExists());
+    Put put = buildable.build();
+
+    try {
+      if (!jdbcService.put(put, connection)) {
+        throw new CrudConflictException(
+            CoreError.JDBC_TRANSACTION_CONFLICT_OCCURRED_IN_INSERT.buildMessage(), txId);
+      }
+    } catch (SQLException e) {
+      throw createCrudException(
+          e, CoreError.JDBC_TRANSACTION_INSERT_OPERATION_FAILED.buildMessage());
+    } catch (ExecutionException e) {
+      throw new CrudException(
+          CoreError.JDBC_TRANSACTION_INSERT_OPERATION_FAILED.buildMessage(), e, txId);
+    }
+  }
+
+  @Override
+  public void upsert(Upsert upsert) throws CrudException {
+    upsert = copyAndSetTargetToIfNot(upsert);
+
+    PutBuilder.Buildable buildable =
+        Put.newBuilder()
+            .namespace(upsert.forNamespace().orElse(null))
+            .table(upsert.forTable().orElse(null))
+            .partitionKey(upsert.getPartitionKey());
+    upsert.getClusteringKey().ifPresent(buildable::clusteringKey);
+    upsert.getColumns().values().forEach(buildable::value);
+    Put put = buildable.build();
```

Comment:
For Upsert in JDBC transactions, we convert the Upsert object to the Put object.